### PR TITLE
Export plugin as an ESM compatible module

### PR DESCRIPTION
**Pull request by santam85**
Avoids getting errors like:
`Uncaught TypeError: helpers$1.merge is not a function`
when loading the plugin as a module.

### CODE REVIEW

**Comment by santam85** [comment]
@simonbrunel if you like this PR, and decide to merge it, could you consider releasing the target branch under the `next` label, in order to enable users of [ng2-charts](https://github.com/valor-software/ng2-charts/pull/1276/commits) to start testing a v3.0 compatible beta?

**Review by simonbrunel — CHANGES_REQUESTED** [review]
@santam85 Thanks you for your help in migrating this plugin. 

I can't guarantee that I will release a beta right after but I will try to do my best to have it released soon. The documentation will also need to be updated to reflect Chart.js 3 changes about registering plugins.

File: docs/.vuepress/plugins/chart-editor/components/ChartEditor.vue
Comment (etimberg) [reply docs/.vuepress/plugins/chart-editor/components/ChartEditor.vue:25]: I created https://github.com/chartjs/Chart.js/pull/8419 as a potential solution to this problem. What do you think?

Comment (LeeLenaleee) [reply docs/.vuepress/plugins/chart-editor/components/ChartEditor.vue:25]: In my migration to V3 I used this code to make sure I had everything imported, dont know if its easyer to read but its less lines:

```
import * as ChartJs from 'chart.js';

ChartJs.Chart.register.apply(null, Object.values(ChartJs).filter((chartClass) => (chartClass.id)));
```

Comment (simonbrunel) [reply docs/.vuepress/plugins/chart-editor/components/ChartEditor.vue:25]: Thanks @etimberg!

@LeeLenaleee It's definitely shorter but a bit hacky. It's also not very robust since it will try to register anything with an `id`, which could be something not registrable. I don't think we should promote this method to Chart.js users.

Comment (santam85) [reply docs/.vuepress/plugins/chart-editor/components/ChartEditor.vue:25]: Thanks, will address once https://github.com/chartjs/Chart.js/pull/8425/files is released (beta.11?)

File: docs/.vuepress/plugins/chart-editor/utils.js
Comment (santam85) [reply docs/.vuepress/plugins/chart-editor/utils.js:30]: It is in a way: the docs present an opportunity to test the ESM version of the lib, and this change is needed in order to use that.

File: package.json
Comment (santam85) [reply package.json:8]: AFAIK bower does not support separate entry points for ESM and UMD. Do you want me to switch the bower.json to always use the ESM?

Comment (simonbrunel) [reply package.json:8]: I thought it was supported, let's not change it then.

File: docs/.vuepress/plugins/chart-editor/utils.js
Comment (simonbrunel) [reply docs/.vuepress/plugins/chart-editor/utils.js:30]: I didn't get that `this` was undefined in the context of an ES module. I will also convert that file on `master` when this PR will be merged.

**Comment by santam85** [comment]
@simonbrunel latest beta broke several tests, any clue what the cause might be? Can't link these errors to any of the listed breaking changes...

**Comment by kurkle** [comment]
@santam85 its the proxied options that default to scriptable being true, and thus tries to resolve the event handlers etc.

Adding something like that to the plugin should help (from #221):
```js
  descriptors: {
    _scriptable: function(name) {
      return !['formatter', 'enter', 'leave', 'click'].includes(name);
    },
  },
```

**Comment by simonbrunel** [comment]
@santam85 This plugin doesn't relying on proxied options (and I'm not sure I would like to) so it's probably better to fully disable this feature for now (`_scriptable: false` if supported) and keep the explicit resolve calls.

@kurkle Shouldn't proxied options for plugins default to `false`? The current behavior seems quite error prone and hard to debug and will probably break a lot of existing plugins.

**Comment by kurkle** [comment]
@simonbrunel yes, it should default to false for plugins.

**Comment by kurkle** [comment]
So, to make this work with beta.11 you'll need to add this:
```js
  descriptors: {
    _scriptable: false,
    listeners: {
      _scriptable: false
    }
  },
```

File: src/plugin.js
Comment (LeeLenaleee) [reply src/plugin.js:147]: It can be removed since that PR is merged already

Comment (santam85) [reply src/plugin.js:149]: But not released yet. I'll remove it when upgrading to next beta

Comment (simonbrunel) [reply src/plugin.js:149]: @LeeLenaleee :) It wasn't last time I checked - but actually the comment should be "when 3.0.0-beta.12 is released". But it doesn't matter, as long as there is a comment that keeps track of the reason of this `descriptors` config.

**Review by simonbrunel — DISMISSED** [review]

File: src/plugin.js
Comment (LeeLenaleee) [reply src/plugin.js:149]: 👍

**Comment by simonbrunel** [comment]
@santam85 A few samples are broken compared to the `chartjs-next` branch:

- Charts/Bar: [`chartjs-next`](https://chartjs-next--chartjs-plugin-datalabels.netlify.app/samples/charts/bar.html) | [PR](https://deploy-preview-222--chartjs-plugin-datalabels.netlify.app/samples/charts/bar.html)
- Events/Highlight: [`chartjs-next`](https://chartjs-next--chartjs-plugin-datalabels.netlify.app/samples/events/highlight.html) | [PR](https://deploy-preview-222--chartjs-plugin-datalabels.netlify.app/samples/events/highlight.html)
- Events/Highlight: [`chartjs-next`](https://chartjs-next--chartjs-plugin-datalabels.netlify.app/samples/advanced/custom-labels.html) | [PR](https://deploy-preview-222--chartjs-plugin-datalabels.netlify.app/samples/advanced/custom-labels.html)
- Maybe more...

File: docs/.vuepress/plugins/chart-editor/components/ChartEditor.vue
Comment (simonbrunel) [inline docs/.vuepress/plugins/chart-editor/components/ChartEditor.vue:27]: I suspect `Chart.defaults` to not contain the plugin defaults yet (which is registered in `ChartView`), thus the broken examples (thought, I'm not sure why it works before). We could try to move the plugin registration here, right after line 25. Could also be related to the proxied options, but I'm not sure if defaults are proxied.

Comment (santam85) [inline docs/.vuepress/plugins/chart-editor/components/ChartEditor.vue:27]: @simonbrunel Having a breakpoint there, it seems the plugin defaults are present even if I don't move the plugin registration. Moving it to the ChartEditor component doesn't fix the problem. Looks to me you are on the right path though, it seems that the config is not containing the default values when rendering the labels.

Comment (kurkle) [reply docs/.vuepress/plugins/chart-editor/components/ChartEditor.vue:27]: The plugin is not registered at that point?

Comment (santam85) [reply docs/.vuepress/plugins/chart-editor/components/ChartEditor.vue:27]: I can see it in the registry, looks to me like it is:
Screenshot 2021-02-24 at 09 19 55

Comment (simonbrunel) [reply docs/.vuepress/plugins/chart-editor/components/ChartEditor.vue:27]: It should, from the `ChartView.vue` import.

Comment (kurkle) [reply docs/.vuepress/plugins/chart-editor/components/ChartEditor.vue:27]: Looks like the plugin is registered in the view. So I don't think the `CHART_DEFAULTS` contain `plugins.datalabels` (what I should have asked).

Comment (kurkle) [reply docs/.vuepress/plugins/chart-editor/components/ChartEditor.vue:27]: There are few things changed in the proxied options in core since beta.11, would be interesting to see if those make a difference. I think we can push next release quite soon.

Comment (santam85) [reply docs/.vuepress/plugins/chart-editor/components/ChartEditor.vue:27]: Seems to be fixed if I link the latest Chart.js master. Eagerly awaiting for next beta, as usual 😄

File: types/index.d.ts
Comment (kurkle) [inline types/index.d.ts:5]: This is probably an issue with the types in beta.12

**Review by kurkle — COMMENTED** [review]
Looks like the samples issue persists, I'll take a closer look at it.

File: types/index.d.ts
Comment (simonbrunel) [reply types/index.d.ts:5]: If the generic base type is still needed, maybe the core could export an nicer alias:

```ts
// chart.js
export type AnyArray = unknown[];
```

```ts
// chart-plugin-datalabels
import { AnyArray } from 'chart.js';

declare module 'chart.js' {
  interface ChartDatasetProperties<TType extends ChartType, TData extends AnyArray> {
    //...
}
```

Comment (simonbrunel) [reply types/index.d.ts:5]: Could even be `AnyData` instead of `AnyArray`.

Comment (santam85) [reply types/index.d.ts:5]: I am just matching the definition here to prevent errors. If only array data-types are needed, would `TData extends Array<any>` work?

Comment (simonbrunel) [reply types/index.d.ts:5]: I think it needs to be the exact same definition as in the core repo, so it would not work. Also, I disabled the use of `any` and enforce array type consistency in this project, so it would need to be `unknown[]`.

Comment (simonbrunel) [reply types/index.d.ts:5]: @santam85 I think I misunderstood your previous comment. Are you suggesting to use `Array<any>` in the core repo?

If so, I would not go that way because `any` isn't good for typing, `unknown` is better because it forces a check of the actual type before using it. Using an alias such as `AnyData` would allow to extend this data type later (if needed) without breaking existing integrations (e.g. `type AnyData = AnyArray | AnyObject`) or `type AnyData = (number | string | DataObject)[]`, etc.).

Comment (santam85) [reply types/index.d.ts:5]: Yeah, I meant in the core repo. Sounds good, will adjust accordingly once decided.

File: docs/.vuepress/plugins/chart-editor/components/ChartEditor.vue
Comment (santam85) [reply docs/.vuepress/plugins/chart-editor/components/ChartEditor.vue:27]: Updated to latest beta and removed the descriptors. Tests are passing, but the default values still seem not to be picked up in the docs. Will look into it a bit more, but I'd appreciate any help, not a pro with Vue.js...

Comment (simonbrunel) [reply docs/.vuepress/plugins/chart-editor/components/ChartEditor.vue:27]: I suspect that proxied options are not enumerable, thus not mergeable which would fail [resolving label options](https://github.com/chartjs/chartjs-plugin-datalabels/blob/50bb58e8c90fc32d048009763997a35a7f15d634/src/plugin.js#L28). It's probably not related to the docs but I think unit tests always explicitly set options to test in the chart config and so never fallback to defaults. And for the [`defaults` unit test](https://github.com/chartjs/chartjs-plugin-datalabels/blob/50bb58e8c90fc32d048009763997a35a7f15d634/types/test/options.default.ts), we test before resolving label options, so it works.

Comment (kurkle) [reply docs/.vuepress/plugins/chart-editor/components/ChartEditor.vue:27]: Probably not it, but this part needs to be updated: 
https://github.com/chartjs/chartjs-plugin-datalabels/blob/9586e39f56b9a5c6dc733d9bb4888b6147e90140/docs/.vuepress/config.js#L20-L32

Comment (kurkle) [reply docs/.vuepress/plugins/chart-editor/components/ChartEditor.vue:27]: @simonbrunel is half right. The proxied options are enumerable, but they do not contain the defaults. So the fix is probably to add the defaults in [that merge](https://github.com/chartjs/chartjs-plugin-datalabels/blob/50bb58e8c90fc32d048009763997a35a7f15d634/src/plugin.js#L28)

Its not compatible with Chart.js v2, but it gives the plugin maintainers even more control on how the options should be handled.

Comment (simonbrunel) [reply docs/.vuepress/plugins/chart-editor/components/ChartEditor.vue:27]: It's possible to access the default options from this passed `options` proxied object, but these default values are not enumerable. I don't think it's the correct behavior. For example, accessing `options.display` will return `true/false` from the defaults, but `{...options}` will not contain `display`, which I would expect to be there.

Comment (simonbrunel) [reply docs/.vuepress/plugins/chart-editor/components/ChartEditor.vue:27]: > So the fix is probably to add the defaults in that merge

I don't think that the correct fix. Instead, we should make sure that these default options are enumerable as part of the proxied `options`. It's currently inconsistent between accessing the options property directly and iterating other the object keys. Plugin developers should not have to care about the internal implementation of this `options` object: it should behave as a plain object as if config were merged with defaults.

Comment (kurkle) [reply docs/.vuepress/plugins/chart-editor/components/ChartEditor.vue:27]: Oh, right. Currently `Object.keys` return the keys in the topmost options object, while `Object.getOwnPropertyNames` returns all the inherited keys. The reasoning behind that might not be valid anymore, have to take a look.

Comment (kurkle) [reply docs/.vuepress/plugins/chart-editor/components/ChartEditor.vue:27]: https://github.com/chartjs/Chart.js/pull/8553 should fix this

Comment (santam85) [reply docs/.vuepress/plugins/chart-editor/components/ChartEditor.vue:27]: Tested it with Chart.js master, seems to be working now! Thanks @kurkle & @simonbrunel!

**Comment by simonbrunel** [comment]
Thanks @santam85. 

Samples are still broken: [legend and tooltip should not be visible](https://github.com/chartjs/chartjs-plugin-datalabels/blob/chartjs-next/docs/.vuepress/config.js#L21). It looks like an issue in the core lib where registering a plugin (and probably other types) doesn't preserve existing defaults:

```js
import {Chart, registrables} from 'chart.js';

Chart.defaults.set({
  plugin: {
    legend: {
      display: false;
    }
  }
})

Chart.register(...registrables);

// Chart.defaults.plugin.legend.display === true
```

I will see how to fix the docs after merging this PR.

**Pull request merged**
